src/persist-client/src/internal/gc.rs

```diff
@@ -194,6 +194,39 @@ where
     }
 
     pub async fn gc_and_truncate(machine: &mut Machine<K, V, T, D>, req: GcReq) {
+        // There's also a bulk delete API in s3 if the performance of this
+        // becomes an issue. Maybe make Blob::delete take a list of keys?
+        //
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html
+        async fn delete_all(
+            blob: &(dyn Blob + Send + Sync),
+            keys: impl Iterator<Item = BlobKey>,
+            metrics: &RetryMetrics,
+            span: Span,
+            semaphore: &Semaphore,
+        ) {
+            let futures = FuturesUnordered::new();
+            for key in keys {
+                futures.push(
+                    retry_external(metrics, move || {
+                        let key = key.clone();
```

Comment:
dumb q: do we need this clone?

Reply:
Yeah, unfortunately:
- Just removing the clone would move the key into the inner future, which isn't allowed in the `retry_external` closure. ("cannot move out of `key`, a captured variable in an `FnMut` closure")
- References won't live long enough, since we're dumping everything into a `FuturesUnordered` which lives outside the loop. ("reference doesn't live long enough").

I tried a bunch of permutations on this one in my first pass at it trying to get the number of clones down -- totally possible this is still suboptimal but I think I have exhausted my ability to improve it. 😅

Reply:
nope, you're right! (my eyes were glazing past the retry_external when I said this)